src/iteration.jl

```diff
@@ -20,50 +20,58 @@ end
 Base.eltype(f::File) = Row
 Base.length(f::File) = getrows(f)
 
-@inline function Base.iterate(f::File, st=1)
+function Base.iterate(f::File)
+    cols = getcols(f)
+    (cols == 0 || getrows(f) == 0) && return nothing
+    c = getcolumn(f, 1)
+    if typeof(c) <: LazyArrays.ApplyArray
+        return Row(f, (1, 1, 1, c.args)), (1, 2, 2, c.args)
+    else
+        return Row(f, 1), 2
+    end
+end
+
+@inline function Base.iterate(f::File, st::Int)
     st > length(f) && return nothing
     return Row(f, st), st + 1
 end
 
-@noinline badcolumnerror(name) = throw(ArgumentError("`$name` is not a valid column name"))
-
-@inline function Base.getproperty(row::Row, name::Symbol)
-    f = getfile(row)
-    i = findfirst(x->x===name, getnames(f))
-    i === nothing && badcolumnerror(name)
-    return getcell(f, gettypes(f)[i], i, getrow(row))
+@inline function Base.iterate(f::File, st)
+    st[3] > length(f) && return nothing
+    if st[2] + 1 > length(st[4][st[1]])
+        st1 += 1
+        st2 = 1
+    else
+        st1 = st[1]
+        st2 = st[2]
+    end
+    return Row(f, st), (st1, st2, st + 1, st[4])
 end
 
-@inline Base.getproperty(row::Row, ::Type{T}, col::Int, name::Symbol) where {T} =
-    getcell(getfile(row), T, col, getrow(row))
+@noinline badcolumnerror(name) = throw(ArgumentError("`$name` is not a valid column name"))
 
-# internal method for getting a cell value
-@inline function getcell(f::File, ::Type{T}, col::Int, row::Int) where {T}
-    ind = metaind(row)
-    @inbounds offlen = gettape(f, col)[ind]
-    missingvalue(offlen) && return missing
-    return getvalue(Base.nonmissingtype(T), f, ind, offlen, col)
+@inline function Base.getproperty(row::Row{Int}, name::Symbol)
+    column = getcolumn(getfile(row), name)
```

Comment:
This is where we get to remove a lot of duplicated logic from tables.jl and just index the `CSV.Column` for our `CSV.Row`.